File: project-simple/src/engine/routing.js
// src/engine/routing.js

import { seg_eta, get_wait, TRANSFER_PENALTY } from './eta.js';

export const MAX_COST_MIN = 90.0;

class MinHeap {
  constructor() {
    this.heap = [];
  }
  push(val) {
    this.heap.push(val);
    this.bubbleUp();
  }
  pop() {
    if (this.heap.length === 1) return this.heap.pop();
    const top = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.sinkDown();
    return top;
  }
  isEmpty() { return this.heap.length === 0; }
  bubbleUp() {
    let index = this.heap.length - 1;
    while (index > 0) {
      let parent = Math.floor((index - 1) / 2);
      if (this.heap[parent][0] <= this.heap[index][0]) break;
      let temp = this.heap[parent];
      this.heap[parent] = this.heap[index];
      this.heap[index] = temp;
      index = parent;
    }
  }
  sinkDown() {
    let index = 0;
    const length = this.heap.length;
    while (true) {
      let left = 2 * index + 1;
      let right = 2 * index + 2;
      let smallest = index;
      if (left < length && this.heap[left][0] < this.heap[smallest][0]) smallest = left;
      if (right < length && this.heap[right][0] < this.heap[smallest][0]) smallest = right;
      if (smallest === index) break;
      let temp = this.heap[smallest];
      this.heap[smallest] = this.heap[index];
      this.heap[index] = temp;
      index = smallest;
    }
  }
}

let _cnt = 0;

export function sssp_from_origin(
  origin_stop_id,
  origin_walk_min,
  depart_hour,
  route_to_stop_list,
  route_stop_pos,
  stop_to_route_dirs,
  eta_exact,
  route_avg_eta,
  wait_lookup
) {
  const dist = {};
  const pred = {};

  const heap = new MinHeap();
  // cost, counter, stop, rd, legs
  heap.push([origin_walk_min, _cnt++, origin_stop_id, null, [
    {
      type: "WALK_START",
      to_stop_id: origin_stop_id,
      walk_time_min: origin_walk_min
    }
  ]]);

  while (!heap.isEmpty()) {
    const [cost, _tc, stop, rd, legs] = heap.pop();
    const key = `${stop}|${rd === null ? 'null' : rd}`;

    if (dist[key] !== undefined) continue;
    dist[key] = cost;
    pred[key] = legs;

    if (cost > MAX_COST_MIN) continue;

    const cur_hour = Math.min(Math.floor(depart_hour + cost / 60), 20);

    // Case A: At stop, haven't boarded
    if (rd === null) {
      const next_rds = stop_to_route_dirs[stop] || [];
      for (const next_rd of next_rds) {
        const route_id = next_rd.substring(0, next_rd.lastIndexOf('_'));
        const wait = get_wait(stop, route_id, cur_hour, wait_lookup);
        const new_cost = cost + wait;
        if (new_cost > MAX_COST_MIN) continue;
        
        const new_key = `${stop}|${next_rd}`;
        if (dist[new_key] !== undefined) continue;
        
        const new_legs = [...legs, {
          type: "BOARD",
          stop_id: stop,
          route_dir: next_rd,
          route_id: route_id,
          wait_min: Math.round(wait * 100) / 100,
        }];
        heap.push([new_cost, _cnt++, stop, next_rd, new_legs]);
      }
    } 
    // Case B: On board route rd
    else {
      const stops_in_route = route_to_stop_list[rd] || [];
      const posMap = route_stop_pos[rd] || {};
      const pos = posMap[stop];
      if (pos === undefined) continue;

      let accum = 0.0;
      const board_stop = _get_board_stop(legs);
      let prev = stop;

      for (let i = pos + 1; i < stops_in_route.length; i++) {
        const next_stop = stops_in_route[i];
        const seg_key = `${prev}->${next_stop}`;
        const seg_t = seg_eta(seg_key, rd, eta_exact, route_avg_eta);
        accum += seg_t;
        const new_cost = cost + accum;

        if (new_cost > MAX_COST_MIN) break;

        const ride_leg = {
          type: "BUS",
          route_dir: rd,
          route_id: rd.substring(0, rd.lastIndexOf('_')),
          board_stop_id: board_stop,
          alight_stop_id: next_stop,
          ride_min: Math.round(accum * 100) / 100,
          n_stops: i - pos,
        };

        // ALIGHT Option
        const alight_key = `${next_stop}|null`;
        if (dist[alight_key] === undefined) {
          const alight_legs = [...legs, ride_leg];
          heap.push([new_cost, _cnt++, next_stop, null, alight_legs]);
        }

        // TRANSFER Option (same-stop only)
        const arr_hour = Math.min(Math.floor(depart_hour + new_cost / 60), 20);
        const other_rds = stop_to_route_dirs[next_stop] || [];
        for (const other_rd of other_rds) {
          if (other_rd === rd) continue;
          
          const other_key = `${next_stop}|${other_rd}`;
          if (dist[other_key] !== undefined) continue;
          
          const other_rid = other_rd.substring(0, other_rd.lastIndexOf('_'));
          const wait_next = get_wait(next_stop, other_rid, arr_hour, wait_lookup);
          const transfer_cost = new_cost + wait_next;
          
          if (transfer_cost > MAX_COST_MIN) continue;
          
          const transfer_legs = [...legs, ride_leg, {
            type: "TRANSFER",
            at_stop_id: next_stop,
            from_route_dir: rd,
            to_route_dir: other_rd,
            to_route_id: other_rid,
            wait_min: Math.round(wait_next * 100) / 100,
            penalty_min: TRANSFER_PENALTY,
          }];
          
          heap.push([transfer_cost + TRANSFER_PENALTY, _cnt++, next_stop, other_rd, transfer_legs]);
        }

        prev = next_stop;
      }
    }
  }

  return { dist, pred };
}

export function build_route_to_poi(dist, pred, poi) {
  const dest_stop = poi.nearest_stop_id;
  if (!dest_stop) return null;

  const stop_key = `${dest_stop}|null`;
  if (dist[stop_key] === undefined) return null;

  const transit_cost = dist[stop_key];
  const walk_end_min = parseFloat(poi.walk_time_min || 0);
  const eta_total = Math.round((transit_cost + walk_end_min) * 100) / 100;

  const legs = [...pred[stop_key], {
    type: "WALK_END",
    from_stop_id: dest_stop,
    dest_name: poi.name || "",
    walk_dist_m: Math.round(parseFloat(poi.walk_dist_m || 0) * 10) / 10,
    walk_time_min: Math.round(walk_end_min * 100) / 100,
  }];

  const transfers = legs.filter(l => l.type === "TRANSFER").length;
  
  return {
    eta_total_min: eta_total,
    transfers: transfers,
    route_legs: legs,
  };
}

function _get_board_stop(legs) {
  for (let i = legs.length - 1; i >= 0; i--) {
    if (legs[i].type === "BOARD") {
      return legs[i].stop_id;
    }
  }
  return "";
}

File: deploy_backend/engine/routing.py
"""
engine/routing.py — Single-Source Shortest Path (SSSP) Dijkstra.

Aturan transit (final):
- WALK hanya di awal (user → stop_awal) dan akhir (stop_akhir → dest)
- TRANSFER hanya pada stop_id yang SAMA (same-stop)
- Multi-transfer diizinkan asal selalu same-stop
- Tidak ada walking antar halte saat transit
"""

from __future__ import annotations
import heapq
import itertools
from typing import Optional

_cnt = itertools.count()

from .eta import seg_eta, get_wait, TRANSFER_PENALTY

MAX_COST_MIN = 90.0   # pruning: abaikan path > 90 menit
_counter = itertools.count()  # tiebreaker agar dict tidak dicompare


def sssp_from_origin(
    origin_stop_id: str,
    origin_walk_min: float,
    depart_hour: int,
    route_to_stop_list: dict[str, list[str]],
    route_stop_pos: dict[str, dict[str, int]],
    stop_to_route_dirs: dict[str, list[str]],
    eta_exact: dict[str, float],
    route_avg_eta: dict[str, float],
    wait_lookup: dict,
) -> tuple[dict, dict]:
    """
    Dijkstra SSSP dari origin_stop_id.

    Returns:
      dist  : {(stop_id, route_dir) → min_cost_menit}
      pred  : {(stop_id, route_dir) → list_of_legs}

    Node = (stop_id, route_dir):
      route_dir=None → sedang di halte, belum naik / sudah turun
      route_dir=X    → sedang dalam perjalanan dengan rute X

    Edge types:
      BOARD     : (stop, None)    → (stop, rd)       cost=wait
      RIDE      : (stop_A, rd)    → (stop_B, rd)     cost=Σ seg_eta
      ALIGHT    : (stop_B, rd)    → (stop_B, None)   cost=0
      TRANSFER  : (stop_X, rd_A) → (stop_X, rd_B)   cost=wait_rd_B (same stop!)
    """
    dist: dict[tuple, float] = {}
    pred: dict[tuple, list] = {}

    # State: (cost, stop_id, route_dir, legs_so_far)
    init_key = (origin_stop_id, None)
    heap = [(origin_walk_min, next(_cnt), origin_stop_id, None, [
        {"type": "WALK_START",
         "to_stop_id": origin_stop_id,
         "walk_time_min": origin_walk_min}
    ])]

    while heap:
        cost, _tc, stop, rd, legs = heapq.heappop(heap)
        key = (stop, rd)

        if key in dist:
            continue
        dist[key] = cost
        pred[key] = legs

        if cost > MAX_COST_MIN:
            continue

        cur_hour = min(int(depart_hour + cost / 60), 20)

        # ── Case A: di halte, belum naik rute ─────────────────────────────
        if rd is None:
            for next_rd in stop_to_route_dirs.get(stop, []):
                route_id = next_rd.rsplit("_", 1)[0]
                wait = get_wait(stop, route_id, cur_hour, wait_lookup)
                new_cost = cost + wait
                if new_cost > MAX_COST_MIN:
                    continue
                new_key = (stop, next_rd)
                if new_key in dist:
                    continue
                new_legs = legs + [{
                    "type": "BOARD",
                    "stop_id": stop,
                    "route_dir": next_rd,
                    "route_id": route_id,
                    "wait_min": round(wait, 2),
                }]
                heapq.heappush(heap, (new_cost, next(_cnt), stop, next_rd, new_legs))

        # ── Case B: sedang dalam rute rd ──────────────────────────────────
        else:
            stops_in_route = route_to_stop_list.get(rd, [])
            pos = route_stop_pos.get(rd, {}).get(stop)
            if pos is None:
                continue

            accum = 0.0
            board_stop = _get_board_stop(legs)
            prev = stop

            for next_stop in stops_in_route[pos + 1:]:
                seg_key = f"{prev}->{next_stop}"
                seg_t = seg_eta(seg_key, rd, eta_exact, route_avg_eta)
                accum += seg_t
                new_cost = cost + accum

                if new_cost > MAX_COST_MIN:
                    break

                ride_leg = {
                    "type": "BUS",
                    "route_dir": rd,
                    "route_id": rd.rsplit("_", 1)[0],
                    "board_stop_id": board_stop,
                    "alight_stop_id": next_stop,
                    "ride_min": round(accum, 2),
                    "n_stops": stops_in_route.index(next_stop) - pos,
                }

                # Opsi ALIGHT (turun di next_stop)
                alight_key = (next_stop, None)
                if alight_key not in dist:
                    alight_legs = legs + [ride_leg]
                    heapq.heappush(heap,
                        (new_cost, next(_cnt), next_stop, None, alight_legs))

                # Opsi TRANSFER di next_stop (same-stop only)
                arr_hour = min(int(depart_hour + new_cost / 60), 20)
                for other_rd in stop_to_route_dirs.get(next_stop, []):
                    if other_rd == rd:
                        continue
                    other_key = (next_stop, other_rd)
                    if other_key in dist:
                        continue
                    other_rid = other_rd.rsplit("_", 1)[0]
                    wait_next = get_wait(next_stop, other_rid, arr_hour, wait_lookup)
                    transfer_cost = new_cost + wait_next
                    if transfer_cost > MAX_COST_MIN:
                        continue
                    transfer_legs = legs + [ride_leg, {
                        "type": "TRANSFER",
                        "at_stop_id": next_stop,
                        "from_route_dir": rd,
                        "to_route_dir": other_rd,
                        "to_route_id": other_rid,
                        "wait_min": round(wait_next, 2),
                        "penalty_min": TRANSFER_PENALTY,
                    }]
                    heapq.heappush(heap, (
                        transfer_cost + TRANSFER_PENALTY,
                        next(_cnt), next_stop, other_rd, transfer_legs
                    ))

                prev = next_stop

    return dist, pred


def build_route_to_poi(
    dist: dict,
    pred: dict,
    poi: dict,
) -> Optional[dict]:
    """
    Dari SSSP result, kembalikan rute ke satu POI.
    poi harus punya: nearest_stop_id, walk_time_min, walk_dist_m.
    """
    dest_stop = poi.get("nearest_stop_id")
    if not dest_stop:
        return None

    stop_key = (dest_stop, None)
    if stop_key not in dist:
        return None

    transit_cost = dist[stop_key]
    walk_end_min = float(poi.get("walk_time_min", 0))
    eta_total = round(transit_cost + walk_end_min, 2)

    legs = pred[stop_key] + [{
        "type": "WALK_END",
        "from_stop_id": dest_stop,
        "dest_name": poi.get("name", ""),
        "walk_dist_m": round(float(poi.get("walk_dist_m", 0)), 1),
        "walk_time_min": round(walk_end_min, 2),
    }]

    transfers = sum(1 for l in legs if l["type"] == "TRANSFER")
    return {
        "eta_total_min": eta_total,
        "transfers": transfers,
        "route_legs": legs,
    }


def _get_board_stop(legs: list) -> str:
    for leg in reversed(legs):
        if leg.get("type") == "BOARD":
            return leg["stop_id"]
    return ""
